src/middlewared/middlewared/plugins/reporting.py

```diff
@@ -0,0 +1,951 @@
+    identifier_plugin = False
+    plugin = 'geom_stat'
+    vertical_label = 'Percent'
+
+    def get_rrd_types(self, identifier):
+        return (
+            (f'geom_busy_percent-{identifier}', 'value', None),
+        )
+
+    def get_title(self):
+        return 'Disk Busy ({identifier})'
+
+
+class DiskGeomLatencyPlugin(GeomStatBase, RRDBase):
+
+    geom_stat_name = 'geom_latency'
+    identifier_plugin = False
+    plugin = 'geom_stat'
+    vertical_label = 'Time,msec'
+
+    def get_rrd_types(self, identifier):
+        return (
+            (f'geom_latency-{identifier}', 'read', None),
+            (f'geom_latency-{identifier}', 'write', None),
+            (f'geom_latency-{identifier}', 'delete', None),
+        )
+
+    def get_title(self):
+        return 'Disk Latency ({identifier})'
+
+
+class DiskGeomOpsRWDPlugin(GeomStatBase, RRDBase):
+
+    geom_stat_name = 'geom_ops_rwd'
+    identifier_plugin = False
+    plugin = 'geom_stat'
+    vertical_label = 'Operations/s'
+
+    def get_rrd_types(self, identifier):
+        return (
+            (f'geom_ops_rwd-{identifier}', 'read', None),
+            (f'geom_ops_rwd-{identifier}', 'write', None),
+            (f'geom_ops_rwd-{identifier}', 'delete', None),
+        )
+
+    def get_title(self):
+        return 'Disk Operations detailed ({identifier})'
+
+
+class DiskGeomQueuePlugin(GeomStatBase, RRDBase):
+
+    geom_stat_name = 'geom_queue'
+    identifier_plugin = False
+    plugin = 'geom_stat'
+    vertical_label = 'Requests'
+
+    def get_rrd_types(self, identifier):
+        return (
+            (f'geom_queue-{identifier}', 'length', None),
+        )
+
+    def get_title(self):
+        return 'Pending I/O requests on ({identifier})'
+
+
+class ARCSizePlugin(RRDBase):
+
+    plugin = 'zfs_arc'
+    vertical_label = 'Bytes'
+    rrd_types = (
+        ('cache_size-arc', 'value', None),
+        ('cache_size-L2', 'value', None),
+    )
+
+    def get_title(self):
+        return 'ARC Size'
+
+
+class ARCRatioPlugin(RRDBase):
+
+    plugin = 'zfs_arc'
+    vertical_label = 'Hits (%)'
+    rrd_types = (
+        ('cache_ratio-arc', 'value', '%name%,100,*'),
+        ('cache_ratio-L2', 'value', '%name%,100,*'),
+    )
+
+    def get_title(self):
+        return 'ARC Hit Ratio'
+
+
+class ARCResultPlugin(RRDBase):
+
+    identifier_plugin = False
+    plugin = 'zfs_arc'
+    vertical_label = 'Requests'
+    rrd_data_extra = """
+        CDEF:total=%name_0%,%name_1%,+
+        XPORT:total:total
+    """
+
+    def get_rrd_types(self, identifier):
+        return (
+            (f'cache_result-{identifier}-hit', 'value', '%name%,100,*'),
+            (f'cache_result-{identifier}-miss', 'value', '%name%,100,*'),
+        )
+
+    def get_title(self):
+        return 'ARC Requests ({identifier})'
+
+    def get_identifiers(self):
+        return ('demand_data', 'demand_metadata', 'prefetch_data', 'prefetch_metadata')
+
+
+class NFSStatPlugin(RRDBase):
```

Comment:
This file is huge for no reason, have you considered moving RRD Plugins (or groups of plugins) into their own files?

Reply:
I agree its big but I couldnt think of any directory structure that would make the problem better and not over complicate plugin loading.

Reply:
I don't think directory structure is necessary, just one dir with cpu.py, memory.py, disk.py, misc.py with few plugins in each

Reply:
The problem is where to put it, I dont like mixing plugins of plugins in the top level middlewared directory.

I think we could have plugins load out of directories so we could have:

middlewared/plugins/foo.py
middlewared/plugins/reporting/__init__.py (Here  we have Service defined)
middlewared/plugins/reporting/cpu.py
middlewared/plugins/reporting/memory.py

?

However I wont have time to work on that this week. Can we merge like that and revisit next month?

Reply:
Sure, I'll create a redmine issue then